### PR TITLE
chore(mapDispatchToProps): port to typescript

### PR DESCRIPTION
**Pull request by tony-go**
Hi 👋 

This pull request follow the typescript migration initiative #1737 

Not 100% confident about type accuracy, but it's good start.

### CODE REVIEW

File: src/connect/mapDispatchToProps.ts [hunk 1]
```diff
@@ -1,21 +1,28 @@
...
+  mapDispatchToProps: ActionCreatorsMapObject | ActionCreator<any>
+) {
   return typeof mapDispatchToProps === 'function'
     ? wrapMapToPropsFunc(mapDispatchToProps, 'mapDispatchToProps')
     : undefined
 }
 
-export function whenMapDispatchToPropsIsMissing(mapDispatchToProps) {
+export function whenMapDispatchToPropsIsMissing(
+  mapDispatchToProps: ActionCreatorsMapObject | ActionCreator<any>
```
Comment (markerikson) [inline src/connect/mapDispatchToProps.ts:14]: :hand: Here, it's probably just `undefined`.

File: src/connect/mapDispatchToProps.ts [hunk 2]
```diff
@@ -1,21 +1,28 @@
...
+) {
   return !mapDispatchToProps
     ? wrapMapToPropsConstant((dispatch) => ({ dispatch }))
     : undefined
 }
 
-export function whenMapDispatchToPropsIsObject(mapDispatchToProps) {
+export function whenMapDispatchToPropsIsObject(
+  mapDispatchToProps: ActionCreatorsMapObject | ActionCreator<any>
```
Comment (markerikson) [inline src/connect/mapDispatchToProps.ts:22]: :hand: Here it can only be the object, effectively.

File: src/connect/wrapMapToProps.ts [hunk 3]
```diff
@@ -20,7 +20,13 @@ export function wrapMapToPropsConstant(
   //  could be a dispatch function in some cases (ex: whenMapDispatchToPropsIsMissing)
   //  and a state object in some others (ex: whenMapStateToPropsIsMissing)
   // eslint-disable-next-line no-unused-vars
-  getConstant: (dispatch: Dispatch) => { dispatch?: Dispatch }
+  getConstant: (dispatch: Dispatch) =>
```
Comment (markerikson) [inline src/connect/wrapMapToProps.ts:23]: :hand: This should probably be reworked to be more generic and flexible - maybe like:

`getConstant: <T>(value: T) => ( () => T)`

In other words, whatever you pass in, it returns a function that returns that thing.

**Review by markerikson — CHANGES_REQUESTED** [review]

File: src/connect/wrapMapToProps.ts (hunk 3 above)
Comment (tony-go) [reply src/connect/wrapMapToProps.ts:23]: It doesn't match with the usage of `wrapMapToPropsConstant` made by `whenMapDispatchToPropsIsMissing`:

```ts
 wrapMapToPropsConstant((dispatch: Dispatch) => ({
      dispatch,
 }))
```

File: src/connect/mapDispatchToProps.ts (hunk 1 above)
Comment (tony-go) [reply src/connect/mapDispatchToProps.ts]: Shoul I let the `!mapDispatchToProps` check ?

File: src/connect/mapDispatchToProps.ts [hunk 4]
```diff
@@ -1,21 +1,28 @@
+import { ActionCreator, ActionCreatorsMapObject, Dispatch } from 'redux'
 import bindActionCreators from '../utils/bindActionCreators'
 import { wrapMapToPropsConstant, wrapMapToPropsFunc } from './wrapMapToProps'
 
-export function whenMapDispatchToPropsIsFunction(mapDispatchToProps) {
+export function whenMapDispatchToPropsIsFunction(
+  mapDispatchToProps: ActionCreatorsMapObject | ActionCreator<any>
```
[the remark this answers is not in the document]
Comment (tony-go) [reply src/connect/mapDispatchToProps.ts]: I put a `FixTypeLater`, I'll try to crack it today.

File: src/connect/wrapMapToProps.ts (hunk 3 above)
Comment (markerikson) [reply src/connect/wrapMapToProps.ts:23]: Good point, it's not just a passthrough

**Comment by markerikson** [comment]
This is probably good enough for the moment. Thanks!

**Pull request merged**